### ISSUE TITLE
timeout does not support onErrorContinue

### ISSUE THREAD

**Issue by DareUrDream** [reactor/reactor-core#1670]
In the below pipe, I would like to timeout the records whose grouping is not complete in X seconds. But timeout does not support onErrorContinue. Also we cannot use `doOnDiscard` as they are additive in nature.

It would be great if have a function to consume timed-out records.

```
                ds.getPublisher()
			.onBackpressureBuffer(15000)
			.groupBy(record -> record.getGroupId())
			.flatMap(group -> group
					.timeout(Duration.ofSeconds(60))
					.bufferUntil(record -> isGroupComplete(record))
					.bufferTimeout(100, Duration.ofSeconds(5))
					.map(listOflistOfRecords -> listOflistOfRecords.stream().flatMap(List::stream).collect(Collectors.toList()))
					.onErrorContinue((th, records) -> {
						LOGGER.error("Timedout records: {}", records);
                                                // TAKE ACTION ON THE RECORDS
					}))
			.filter(records -> {
				return (publishController.shouldIPublish()) ? true 
						: records.get(0).getCreatedTimestamp() <= (publishController.stopRequestTimestamp() - 5);
			})
			.doOnDiscard(List.class, records -> {
				if(! records.isEmpty()) {
					LOGGER.error("Discarded: {}", records);
					discardedRecords.put(records, new Object());
				} else {
					LOGGER.error("Empty record received. This should never happen.");
				}
			})
			.map(record -> Collections.unmodifiableList(Enricher.enrich(record)))
			.map(dbRecords -> RecordTransformer.transform(dbRecords))
			.retryBackoff(MAX_RETRY, Duration.ofSeconds(FIRST_BACKOFF_IN_SECONDS), Duration.ofSeconds(MAX_BACKOFF_IN_SECONDS))
			.publishOn(Schedulers.single());
```

Cheers,
DareUrDream

**Comment by simonbasle** [comment]
This is definitely not a direction we want to go for `onErrorContinue`, which is only there to remove the terminal aspect of transient errors.

`doOnDiscard` is a better candidate, although I'm not sure what the issue with its additive nature is...

I'm not sure either that I understand what you expect of the `timeout(60s)`, but here might be an improvement: if you put the `timeout` at the end of the `flatMap`, early buffers will pass through and if the whole group takes more than 60s to process, the **last** buffer will be discarded. `doOnDiscard` can then pick that last buffer and process the items that were the most late (in the 55s-60s range since you have a `bufferTimeout(100, 5s)`).

Would that help?

**Comment by DareUrDream** [comment]
`group.timeout(Duration.ofSeconds(60))` would help me to close the buffer if not completed (based on `bufferUntil(record -> isGroupComplete(record))` in 60 seconds...But then when the timeout exception is thrown I would like to take the records that were there in the buffer an take and action on them. 

`doOnDiscard` is something that I want to be executed only for the filtered records. So the moment I add another `doOnDiscard` the additive nature creates an issue for me.

So basically what am looking for is to consume the records that were there in the group when the timeout exception is thrown

**Comment by simonbasle** [comment]
ah I remember now you chimed in in https://github.com/reactor/reactor-core/issues/821. Is that still the same use case? (cutting the buffer at the 60s mark, and continuing a new buffer after that?)

**Comment by DareUrDream** [comment]
Yes.

To unblock we moved ahead and defined this flow definition..

**Comment by simonbasle** [comment]
would that approach work: capturing a timestamp that reflects the age of the elements in the group and making that part of the predicate of `bufferUntil`:

```java
.flatMap(group -> {
    AtomicLong first = new AtomicLong(-1);
    return group.timestamp()
        .bufferUntil(tuple2 -> {
            Long timestamp = tuple2.getT1();
            Record record = tuple2.getT2();
			first.compareAndSet(-1, timestamp);
            Long age = timestamp - first.get();
            if (isGroupComplete(record) || age >= 60_000) {
                first.set(-1);
                return true;
            }
            return false;
       })
        //... at some point unwrap the Tuple2 back into the value
       ;
})
```

**Comment by DareUrDream** [comment]
Timeout was helping with the critical information that this group is still not complete and hence some action is taken when that exception was thrown. How would I differentiate and take action ?

Is there no simple way to peep into the records when that exception is thrown. How about our own custom exception which contains the data as well ?

**Comment by simonbasle** [comment]
`timeout` operator doesn't know about the group or the fact that there is buffering involved, it simply cancels its source, so it cannot extract data from that source...

And then, how would `bufferTimeoutUntil` have helped if it was available anyway? (on the same model as `bufferTimeout`, which doesn't differentiate between a buffer that is cut out of size or out of timeout).

I'm trying to come up with different angles to your problem, but I think this is way too specific for a generic operator to be possible...

**Comment by simonbasle** [comment]
re the "how would I differentiate": how about when you get a buffer you check the last `record` in it and apply the `isGroupComplete(record)` predicate again? unless that is not idempotent...

**Comment by DareUrDream** [comment]
`timeout` - Am not sure of the generic use-cases but I guess it would always help the exception carries the current source data, so that consumers can take the call.

`bufferTimeoutUntil` -- This would have helped write more concise code -- Either group is complete or the max time has expired. In this case, I was hoping to get an tuple carrying the exception if any and source data (irrespective of exception ).

Yes, I can re-evaluate `isGroupComplete(record)` but then want to avoid it as it is a CPU intensive activity.

**Comment by DareUrDream** [comment]
@simonbasle  any updates ??

**Comment by simonbasle** [comment]
@DareUrDream `bufferTimeoutUntil` isn't trivial to implement. We might get to it at some point, but this isn't top priority right now.

Discarding, which is definitely the way to go in your case (not supporting onErrorContinue on timeout, which is unrelated). I don't think the `Tuple2<Exception, UnfinishedBuffer>` is ever going to happen.

That said, putting a `doOnDiscard` *inside* a `flatMap` makes that discard handler scoped to the flatMap. It can still receive elements from upstream (including unrelated types), but cancellation of the `group` due to putting a `timeout` on it inside the `flatMap` should only invoke the `doOnDiscard` handler of that particular `flatMap` inner sequence.

So maybe a lower hanging fruit is ensuring that groups of `groupBy` use the `doOnDiscard` handlers.

This can be explored separately, closing this as duplicate of the bufferTimeoutUntil issue.

**Comment by simonbasle** [comment]
This is a duplicate of #821

**Comment by DareUrDream** [comment]
@simonbasle I know this issue is closed but am being a bit lazy here and asking you the below... My above data flow definition is starting to fall in load scenario's as I am forming thousands of groups per minute and they stay there for sometime...Any suggestions here to get around this problem.

Thanks a ton !!
